docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+- During `Bridge` definitions, type definition mistakes in remote facades become more visible. In these cases, be aware of the cascading effect of changing/restricting an argument type in public API when you consider such changes.
+
+**Example**
+
+```php
+namespace Spryker\Zed\Product\Business;
+
+interface ProductFacadeInterface
+{
+public function addProduct(ProductAbstractTransfer $productAbstractTransfer, array $productConcreteCollection): int;
+```
+
+```php
+namespace Spryker\Zed\ProductApi\Dependency\Facade;
+
+class ProductApiToProductBridge implements ProductApiToProductInterface
+{
+  /**
+   * @var \Spryker\Zed\Product\Business\ProductFacadeInterface
+   */
+  protected $productFacade;
+
+  public function __construct($productFacade)
+  {
+    $this->productFacade = $productFacade;
+  }
+
+  public function addProduct(ProductAbstractTransfer $productAbstractTransfer, array $productConcreteCollection): int
+  {
+    return $this->productFacade->addProduct($productAbstractTransfer, $productConcreteCollection);
+  }
+}
+```
+
+## Plugin
+
+```
+[Organization]
+├── Zed
+│   ├── [Module]
+│   │    └── Communication
+│   │        └── Plugin
+│   │            └── [ConsumerModule]
+│   │                └── [PluginName]Plugin.php
+│   └── [ConsumerModule]Extension
+│       └── Dependency
+│           └── Plugin
+│               └── [PluginInterfaceName]PluginInterface.php
+│
+├── Yves
+│   ├── [Module]
+│   │   └── Plugin
+│   │       └── [ConsumerModule]
+│   │           └── [PluginName]Plugin.php
+│   └── [ConsumerModule]Extension
+│       └── Dependency
+│           └── Plugin
+│               └── [PluginInterfaceName]PluginInterface.php
+│
+├── Glue
+│   ├── [Module]
+│   │   └── Plugin
+│   │       └── [ConsumerModule]
+│   │           └── [PluginName]Plugin.php    
+│   └── [ConsumerModule]Extension
+│       └── Dependency
+│           └── Plugin
+│               └── [PluginInterfaceName]PluginInterface.php
+│
+└── Client
+    ├── [Module]        
+    │   └── Plugin
+    │       └── [ConsumerModule]
+    │           └── [PluginName]Plugin.php    
+    └── [ConsumerModule]Extension
+        └── Dependency
+            └── Plugin
+                └── [PluginInterfaceName]PluginInterface.php
+
+```
+
+**Description**
+- `Plugins` are classes which are used to realize Inversion-Of-Control - instead of a direct call to another module's [facade](#facade-design-pattern), a `plugin` can be provided as an optional and configurable class.
+- `Plugins` are the only classes that can be instantiated across modules.
+- See more details about the implementation in [Plugin Interface](#plugin-interface) component, and instantiation in [Dependency Provider](#dependency-provider).
+
+**Conventions**
+- The `Plugin` MUST NOT contain business logic but delegates to the underlying [facade](#facade-design-pattern).
```

Comment:
This rule adds a lot of overhead in the code and generates a lot of methods in the Facade for only one case (for one specific plugin). I vote for removing it or making it less strict.

For project Plugins the rule 100% must not be strict.

Reply:
I like the idea, but from real life experience I can tell you it brings a lot of mess, especially for projects:
when plugin contains the logic, then tests are written against the plugin, and it gets wierd:
some plugins are tested , since they have business logic inside
some plugins are not tested, since the facade method they call is tested.
some plugins are tested with just a single facade method, and the facade method tested separately second time.

so my suggestion for projects is:
```suggestion
- The `Plugin` MAY contain business logic.
```

Reply:
Plugins are core concept. It is not recommended for projects to use Plugins at all. Therefore, nothing applies anyways.

Reply:
@matweew Although I'm fully on your side, we tried to remove this rule for many times and it triggers a huge debate. I think we can't remove it atm without aligning.